### PR TITLE
Core/Porting OMP Parallel for loops to Parallel Utilities_15

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
This PR ports the OMP Parallel Loops to Parallel Utilities

-ParallelUtilities

**Changelog**
-Replaced Simple OMP For loop with Parallel Utilities - IndexPartition()
-Replaced OMP For loops with Private variables with Parallel Utilities - IndexPartition()
-Replaced OMP Parallel For loops with Private variables with ParallelUtilities(TLS)
-Still few loops from the same file need to port after clarification from PR #8803

### CODE REVIEW

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h [hunk 1]
```diff
@@ -749,10 +750,11 @@ class ResidualBasedEliminationBuilderAndSolverSlip
         }
 
         //calculate the total size of the system
-        int total_size = 0.0;
-        #pragma omp parallel for reduction(+:total_size)
-        for (int i = 0; i < number_of_threads; i++)
-            total_size += local_sizes[i];
+        int total_size = 0;
+
+        total_size = IndexPartition<std::size_t>(number_of_threads).for_each<SumReduction<int>>([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h]:
```suggestion
        std::size_t total_size = IndexPartition<std::size_t>(number_of_threads).for_each<SumReduction<std::size_t>>([&](std::size_t Index){
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h [hunk 2]
```diff
@@ -2149,10 +2144,9 @@ class ResidualBasedEliminationBuilderAndSolverWithConstraints
 
         // Filling constant vector
         if (ComputeConstantVector) {
-            #pragma omp parallel for
-            for (int i = 0; i < static_cast<int>(BaseType::mEquationSystemSize); ++i) {
-                rConstantVector[i] = 0.0;
-            }
+            IndexPartition<std::size_t>(BaseType::mEquationSystemSize).for_each([&rConstantVector](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: Gcc 4.8.4 seems to have a problem here
~~~
/__w/Kratos/Kratos/kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h: In member function 'bool Kratos::ResidualBasedEliminationBuilderAndSolverWithConstraints<TSparseSpace, TDenseSpace, TLinearSolver>::ComputeConstraintContribution(typename Kratos::ResidualBasedEliminationBuilderAndSolverWithConstraints<TSparseSpace, TDenseSpace, TLinearSolver>::BaseType::TSchemeType::Pointer, Kratos::ModelPart&, bool, bool)':
/__w/Kratos/Kratos/kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h:2147:40: error: invalid declaration of 'Kratos::ResidualBasedEliminationBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver>::mEquationSystemSize'
             IndexPartition<std::size_t>(BaseType::mEquationSystemSize).for_each([&rConstantVector](std::size_t Index){
~~~

```suggestion
            IndexPartition<std::size_t>(mEquationSystemSize).for_each([&rConstantVector](std::size_t Index){
```

does this work @ashishdarekar? Can you first try locally?
If not then you could try with a temporary variable

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h [hunk 3]
```diff
@@ -2109,18 +2104,18 @@ class ResidualBasedEliminationBuilderAndSolverWithConstraints
...
             // NOTE: dofs are assumed to be numbered consecutively
             const auto it_dof_begin = BaseType::mDofSet.begin();
-            #pragma omp parallel for
-            for(int k = 0; k < static_cast<int>(mDoFToSolveSystemSize); ++k) {
-                auto it_dof = it_dof_begin + k;
-                if (k < static_cast<int>(BaseType::mEquationSystemSize)) {
+
+            IndexPartition<std::size_t>(mDoFToSolveSystemSize).for_each([&](std::size_t Index){
+                auto it_dof = it_dof_begin + Index;
+                if (Index < BaseType::mEquationSystemSize) {
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: seems like you need to explicitly capture `mEquationSystemSize` as it is protected and hence cannot be accessed in the lambda without being captured

I am confused as to whether this is a bug in Gcc 4.8 or not

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h (hunk 2 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: This is not working, `error: use of undeclared identifier 'mEquationSystemSize'`

Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]:
```suggestion
            IndexPartition<std::size_t>(BuilderAndSolverBaseType::mEquationSystemSize).for_each([&rConstantVector](std::size_t Index){
```

can you try with this?

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h (hunk 3 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: How should I capture it again ??  `[&]` includes everything right ?

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: I don't think `&` captures the class instance itself (`this`)
Didn't check the docs though

Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: Ohh I see... I will try with `this`

Comment (rubenzorrilla) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: `&` captures everything in the scope where the lambda is declared. To capture the class itself you need to specify the `this`.

Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: Still the error is same ,
`>::mEquationSystemSize' is protected
     unsigned int mEquationSystemSize; /// Number of degrees of freedom of the problem to be solve
                  ^
In file included from /__w/Kratos/Kratos/kratos/tests/cpp_tests/strategies/builder_and_solvers/test_builder_and_solver.cpp:44:0,
                 from /__w/Kratos/Kratos/build/Custom/kratos/CMakeFiles/KratosCore.dir/Unity/unity_8_cxx.cxx:8:
/__w/Kratos/Kratos/kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h:2110:27: error: within this context
                 if (Index < BaseType::mEquationSystemSize)`

Am I making some mistake again?

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: hmmmm
can you try this?
```suggestion
                if (Index < this->mEquationSystemSize) {
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h (hunk 2 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: Its working fine on local machine, 
but on CentOS giving error :
`ool)':
/__w/Kratos/Kratos/kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h:2147:40: error: invalid declaration of 'Kratos::BuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver>::mEquationSystemSize'
             IndexPartition<std::size_t>(BuilderAndSolverBaseType::mEquationSystemSize).for_each([&rConstantVector](std::size_t Index){`

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: I saw

as a last test, can you try with `this->`?

**Comment by philbucher** [comment]
you missed one but I think it is working!

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h (hunk 3 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: Working now . Thanks :)

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h (hunk 2 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: okay ..

Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h]: worked !!!!Thanks :)

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h [hunk 4]
```diff
@@ -749,10 +750,9 @@ class ResidualBasedEliminationBuilderAndSolverSlip
         }
 
         //calculate the total size of the system
-        int total_size = 0.0;
-        #pragma omp parallel for reduction(+:total_size)
-        for (int i = 0; i < number_of_threads; i++)
-            total_size += local_sizes[i];
+        std::size_t total_size = IndexPartition<std::size_t>(number_of_threads).for_each<SumReduction<int>>([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h:753]:
```suggestion
        std::size_t total_size = IndexPartition<std::size_t>(number_of_threads).for_each<SumReduction<std::size_t>>([&](std::size_t Index){
```

**Review by philbucher — APPROVED** [review]
👍 

Thx

**Pull request merged**